Sources/SmokeAWSMetrics/CloudWatchPendingMetricsQueue.swift

```diff
@@ -0,0 +1,285 @@
+// Copyright 2018-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License").
+// You may not use this file except in compliance with the License.
+// A copy of the License is located at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+//
+//  CloudWatchPendingMetricsQueue.swift
+//  SmokeAWSMetrics
+//
+
+import Foundation
+import CloudWatchClient
+import CloudWatchModel
+import Logging
+
+private func iso8601DateFormatter() -> DateFormatter {
+    let formatter = DateFormatter()
+    formatter.calendar = Calendar(identifier: .iso8601)
+    formatter.locale = Locale(identifier: "en_US_POSIX")
+    formatter.timeZone = TimeZone(secondsFromGMT: 0)
+    formatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss.SSSXXXXX"
+    return formatter
+}
+
+internal extension Date {
+    var iso8601: String {
+        return iso8601DateFormatter().string(from: self)
+    }
+}
+
+internal enum CloudWatchPendingMetricsQueueError: Error {
+    case shutdownAttemptOnUnstartedQueue
+}
+
+private struct Entry {
+    let namespace: String
+    let data: MetricDatum
+}
+
+internal class CloudWatchPendingMetricsQueue {
+    private var pendingEntries: [Entry]
+    private let logger: Logger
+    private let cloudWatchClient: CloudWatchClientProtocol
+    private let shutdownDispatchGroup: DispatchGroup
+    
+    private let accessQueue = DispatchQueue(label: "com.amazon.SmokeAWSMetrics.CloudWatchPendingMetricsQueue.accessQueue")
+    private let queueConsumingTaskIntervalInSeconds = 2
+    private let maximumDataumsPerRequest = 100
```

Comment:
According to https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/cloudwatch_limits.html, PutMetricsData only supports up to 20 MetricDatum items per request.

Reply:
Good catch, thanks. I have changed the limit with an explanation to that document.